PR for waypoint_updater and twist_controller update

chriskcheung opened the pull request:
Hi Team, I updated waypoint_updater and twist_controller to handle deceleration base on traffic light and obstacle detection topics. Please review my changes and let me know if I missed anything or recommendation for improvement. Thanks.

chriskcheung committed:
Updated waypoint_updater and twist_controller

- Added subscriber to traffic_waypoint and obstacle_waypoint topics
- Added get_final_lane to pick the closest waypoint upon traffic light and obstacle detection
- Added decelerate_waypoints to update base_waypoints' linear velocity, x
- update publish_waypoints to use get_final_lane
- updated traffic_callback and obstacle_callback accordingly
- updated brake torque from 400 to 700Nm for Carla's braking per lecture recommendation

soniclue commented on ros/src/waypoint_updater/waypoint_updater.py:
```diff
@@ -40,6 +42,8 @@ def __init__(self):
         rospy.Subscriber('/current_pose', PoseStamped, self.pose_callback)
         rospy.Subscriber('/base_waypoints', Lane, self.base_waypoints_callback)
         # TODO: Add a subscriber for /traffic_waypoint and /obstacle_waypoint below
```
remove?

soniclue commented on ros/src/waypoint_updater/waypoint_updater.py:
```diff
@@ -92,11 +98,67 @@ def publish_waypoints(self, index):
+        lane = self.get_final_lane(index)
         lane.header = self.base_waypoints.header
-        lane.waypoints = self.base_waypoints.waypoints[index:index + _NUM_WAYPOINTS_AHEAD]
+        #lane.waypoints = self.base_waypoints.waypoints[index:index + _NUM_WAYPOINTS_AHEAD]
```
remove?

chriskcheung replied:
this is replaced with get_final_lane(index). I will clean up the commented line and resubmit.

soniclue commented on ros/src/waypoint_updater/waypoint_updater.py:
```diff
@@ -92,11 +98,67 @@ def publish_waypoints(self, index):
+        #lane.waypoints = self.base_waypoints.waypoints[index:index + _NUM_WAYPOINTS_AHEAD]
         self.final_waypoints_pub.publish(lane)
 
+    # update final Lane's waypoints base on traffic light or obstacle waypoint index
```
PEP8 style of function docstring.

chriskcheung replied:
Will update it following PEP8 docstring

soniclue commented on ros/src/waypoint_updater/waypoint_updater.py:
```diff
@@ -92,11 +98,67 @@ def publish_waypoints(self, index):
+            lane.waypoints = self.declerate_waypoints(base_waypoints, closest_wp_idx, target_wp_idx)
+        return lane
+
+    # find decelerating waypoints
```
same here

chriskcheung replied:
Noted

soniclue commented on ros/src/waypoint_updater/waypoint_updater.py:
```diff
@@ -116,13 +178,13 @@ def base_waypoints_callback(self, base_waypoints):
         rospy.loginfo('base_waypoints initialized')
 
-    def traffic_cb(self, msg):
+    def traffic_callback(self, data):
```
same here and below.

chriskcheung replied:
Noted

soniclue reviewed (approved):
just some code style stuff.

chriskcheung committed:
Clean up code base on review

- added PEP8 style of function docstring for function headers
- moved stop_idx outside of base_waypoints for loop
- added comment to deceleration forumula used

chriskcheung reviewed (commented):
I move this out of base_waypoint for loop as it is static from input parameters
